Sequence functionality: API revisions

dhardy opened the pull request:
Lets start trying to implement https://github.com/dhardy/rand/issues/82 !

The API is not fixed but has been discussed in the above thread. Comments welcome (but please read the above first).

@pitdicker this is an upstream branch; do what you like with it!

dhardy commented:
I'm currently pushing some code for this.

I should note that this is not intended for merging within the 0.5 series, so will probably not be merged soon. At some point we should create a 0.5 branch and update master for 0.6, but not yet.

dhardy commented:
That completes the basic implementations. Still to do:

- API critique / bikeshed — may be worth opening a new issue to discuss this
- revise `sample_indices` somehow, probably (could be separate PR)
- Is returning an iterator for `choose_multiple` appropriate? Is `choose_multiple(rng, m).collect()` as fast as if we just output a vector?
- any other algorithm adjustments?
- module documentation
- clean up / rebase

dhardy commented:
Anyone know why rustdoc can't find `Index`?

vks commented:
> Update minimum Rustc version to 1.26

Is this planned for Rand 0.6?

Instead of having `choose_multiple*`, couldn't we just provide an iterator that can be `collect`ed? (Would be nicer in cases where the size of the iterator is known.)

dhardy commented:
Yes, the "remove deprecations" thing will probably become it's own PR — but not before we have a 0.5 branch, and I'm not sure we should do that yet.

pitdicker commented:
We should probably be brave and have a couple of point releases first.

dhardy commented:
TODO: benchmark whether `partial_shuffle` is actually an improvement over `choose_multiple` (which does in fact yield shuffled results).

sicking commented:
I didn't see this PR existed before adding #490. There's some overlap (mainly in where the implementation of the `choose` methods should live), but so far they mainly look independent.

dhardy commented:
@sicking would you please review this PR in more detail since it is close to your own #490? The big differences I see:

- Weighted choices: I haven't done that yet; hopefully it can be added into this API somehow.
- Methods on the `Rng` trait: IMO these methods should never be more than a wrapper around something else; I know @vks thinks most shouldn't even be there and I believe @pitdicker views these mostly as convenience methods. I guess we could keep the existing methods (`choose`, `choose_mut` and `shuffle`) as wrappers instead of deprecating (as in this PR), but I'm not keen on adding several new things there as your PR does.
- Trait names: `SliceRandom` vs `SliceExt`, `IteratorRandom` vs `IteratorExt`: I don't have much opinion on this, though `use rand::seq::SliceRandom;` seems redundant. I don't know if `SliceExt` would be confusing or conflict for users?
- This PR has a few extras methods on the traits: `shuffle`, `choose_multiple`, `partial_shuffle`

If some parts of this/your PR are contentious but we can find an agreeable sub-set that would also help make progress.

sicking commented on src/seq.rs:
```diff
@@ -9,61 +9,337 @@
+            None
+        } else {
+            let len = self.len();
+            Some(&mut self[rng.gen_range(0, len)])
```
Put `self.len()` inline like in the previous function?

dhardy replied:
Causes a failure with rustc 1.22

sicking commented on src/seq.rs:
```diff
@@ -9,61 +9,337 @@
+/// trait.SliceExt.html#method.choose_multiple).
+#[cfg(feature = "alloc")]
+#[derive(Debug)]
+pub struct SliceChooseIter<'a, S: ?Sized + 'a, T: 'a> {
```
You should also implement ExactSizeIterator for this struct.

dhardy replied:
Done (comment not auto-hidden)

sicking commented on src/seq.rs:
```diff
@@ -9,61 +9,337 @@
+    type Item = &'a T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.indices.next().map(|i| &(*self.slice)[i])
```
You could call `get_unchecked` here to save the bounds-check.

dhardy replied:
The required trait is nightly-only for now: https://doc.rust-lang.org/std/slice/trait.SliceIndex.html#tymethod.get_unchecked

sicking commented on src/seq.rs:
```diff
@@ -225,19 +506,74 @@ fn sample_indices_cache<R>(
+        assert_eq!(v.iter().choose(&mut r), Some(&3));
+
+        let v: &[isize] = &[];
+        assert_eq!(v.choose(&mut r), None);
```
These tests aren't very comprehensive. But I'm happy to add the tests I have in #490 over there.

dhardy replied:
Sure, I'll look forward to it.

sicking commented:
If we really want to take advantage of namespaces and avoid redundancy, maybe we should be so bold as simply calling the traits `rand::Slice` and `rand::Iterator`. But as someone that has lived in non-namespaced languages most of my life, those names definitely make me a little uncomfortable. I don't know if there would be any practical problems with it though?

dhardy commented on src/seq.rs:
```diff
@@ -9,61 +9,337 @@
+        for (i, elem) in self.enumerate() {
+            let k = rng.gen_range(0, i + 1 + amount);
+            if k < amount {
+                buf[k] = elem;
```
I have no idea. How does one check that (other than by benchmark)? I tried looking at the generated code for a small example, but even the MIR is 1200 lines, and likely before optimisations. Assembly is 5800.

dhardy commented on src/seq.rs:
```diff
@@ -9,61 +9,337 @@
+        let mut len = 0;
+        while len < amount {
+            if let Some(elem) = self.next() {
+                buf[len] = elem;
```
I tried this, and benchmarks show no difference. I would guess that the check can get eliminated:
```rust
        for (i, slot) in buf.iter_mut().enumerate() {
            if let Some(elem) = self.next() {
                *slot = elem;
            } else {
                // Iterator exhausted; stop early
                return i;   // length
            }
        }
```

sicking replied:
I personally prefer the approach this approach as it doesn't rely on the optimizer removing the check. But I'm old and crotchety :). Either way is totally fine with me.

dhardy replied:
Same performance, so I'd rather go with what looks nicer :smile:

sicking replied:
Cool

dhardy commented on src/seq.rs:
```diff
@@ -9,61 +9,337 @@
+    /// 
+    /// Complexity is `O(n)` where `n` is the length of the iterator.
+    #[cfg(feature = "alloc")]
+    fn choose_multiple<R>(mut self, rng: &mut R, amount: usize) -> Vec<Self::Item>
```
I tried combining the implementations too. Problem is we want allocated-on-first-use storage which `Vec` gives us for free but isn't compatible with buffers; some kind of trait might be able to abstract over this, but as you said it just results in more code in total.

About the return type, I'm not sure; both have their merits. However what about consistency? I think it would be possible to make `SliceExt::choose_multiple` return a `Result` too, even using iterators as it does now.

sicking replied:
Good point about consistency. Would make sense to return a Result<Iterator, Iterator> for `SliceExt::choose_multiple`.

You could even argue that we should be consistent with `choose` and return an `Result<T, Something>`, but that's more dubious.

I don't have strong feelings about this. Maybe sticking with what we have now makes for the cleanest API.

dhardy replied:
The `std` library already uses `Option` in cases where `None` is kind-of an error. I don't think it makes sense to use `Result<T, ()>` (or `NoInput` instead of `()`).

sicking replied:
Yeah, I was thinking that too. I think just returning a Vec/Iterator might be the cleanest solution.

dhardy commented on src/seq.rs:
```diff
@@ -9,61 +9,337 @@
+    #[cfg(feature = "alloc")]
+    fn choose_multiple<R>(&self, rng: &mut R, amount: usize) -> SliceChooseIter<Self, Self::Item>
+        where R: Rng + ?Sized
+    {
```
This one could also be made to return a `Result`, but it's a bit ugly. Thoughts? Doesn't seem to have much overhead, other than the ugly code.
```rust
    fn choose_multiple<R>(&self, rng: &mut R, amount: usize) ->
        Result<SliceChooseIter<Self, Self::Item>, SliceChooseIter<Self, Self::Item>>
        where R: Rng + ?Sized
    {
        if amount <= self.len() {
            Ok(SliceChooseIter {
                slice: self,
                _phantom: Default::default(),
                indices: sample_indices(rng, self.len(), amount).into_iter(),
            })
        } else {
            let mut indices: Vec<usize> = (0..self.len()).collect();
            indices.shuffle(rng);
            Err(SliceChooseIter {
                slice: self,
                _phantom: Default::default(),
                indices: indices.into_iter(),
            })
        }
    }
```

sicking replied:
The neat-freak in me thinks that maybe we should just go with the old approach.

dhardy replied:
"old" meaning without the `Result`? @pitdicker what do you think (also about using `Result` to indicate not enough elements on the other `choose_multiple` above)?

sicking replied:
"old" meaning as you have it now, not a `Result`. Same with the other `choose_multiple`.

I want to be clear that I really don't have a strong opinion either way. Other than that I think it makes sense to be consistent.

dhardy commented:
@sicking I addressed some of your comments, though not the trait names (TBD).

dhardy commented on src/seq.rs:
```diff
@@ -9,61 +9,337 @@
+            // Continue until the iterator is exhausted
+            for (i, elem) in self.enumerate() {
+                let k = rng.gen_range(0, i + 2);
+                if k == 0 {
```
`gen_ratio` gets a big speed-up:
```
# current
test seq_iter_choose_from_100                ... bench:       1,053 ns/iter (+/- 55)
# using (i + 2) as u32
test seq_iter_choose_from_100                ... bench:         491 ns/iter (+/- 16)
# using TryFrom
test seq_iter_choose_from_100                ... bench:         526 ns/iter (+/- 14)
```
Unfortunately TryFrom is still not stable, and also doesn't support working with `usize` directly (https://github.com/rust-lang/rust/issues/49415). I'm not certain we should use `as` here and because this is an iterator without known exact size we have no choice but to check each time for correctness.

sicking replied:
Yeah. Let's stick with what you have. This might be an argument for making `gen_ratio` generic over type though. But lets worry about that separately.

sicking commented on src/seq.rs:
```diff
@@ -9,61 +9,337 @@
+            i -= 1;
+            // lock element i in place.
+            self.swap(i, rng.gen_range(0, i + 1));
+        }
```
Is the perf difference meaningful? I feel like it's quite surprising to operate on a slice from end, but I guess it could be ok as long as we are very clearly documenting it.

sicking commented:
> I am starting to prefer `sample` over `choose` and `pick`.

I have a weak preference for `pick`. Mainly because I associate it with "without replacement", which makes `pick_multiple` descriptive. OTOH `choose` matches python. Definitely no strong preference either way, so happy to go with any of these three.

sicking commented:
> I have a weak preference for `pick`. Mainly because I associate it with "without replacement", which makes `pick_multiple` descriptive. OTOH `choose` matches python. Definitely no strong preference either way, so happy to go with any of these three.

Actually, I take this back. I have no preference either way for any of the three names :).

dhardy commented:
In my mind, *choose* has an element of decisiveness to it, where *pick* often means *take any without care*. *Sample* is of course often used in stats. Thus it seems intuitive (to me at least) that `rng.choose(list)` or `list.choose(rng)` means take a random element. *Sample* often implies multiple, thus I guess we could have `list.choose(rng)` for one elt and `list.sample(rng, number)` for multiple elements.

> Mainly because I associate it with "without replacement"

On the other hand, that *pick* should mean *without replacement* is not obvious to me! Actually, I would assume any of these terms would imply sampling without replacement unless otherwise specified.

(However, although it would be easy to write something like `(0..n).map(|_| list.choose(&mut rng).clone()).collect()` to sample with replacement, that's not quite optimal due to the missed opportunity to use `Uniform::new` to cache index sampling parameters; i.e. there might be some value in a sampling-with-replacement function. On the other hand, at least it's easy to write a correct and reasonably fast implementation, and a faster version is not hard to write: `let index_range = Uniform::new(0, len); (0..n).map(|_| rng.sample(index_range)).map(|i| list[i].clone()).collect()`. Also I'm not sure that sampling-with-replacement is so commonly used.)

dhardy commented:
> I personally like SliceRandom over SliceExt.

> Adding SliceExt and IteratorExt to the prelude seems like a good idea to me. That also makes the trait names less important.

Both good points. In a way the names don't affect users at all (unless manually imported), *unless they happen to conflict* (e.g. another crate makes its own extension trait on slices and calls that `SliceExt`). So I agree with the rename.

dhardy commented:
> Regarding the trait names. I whatever name that we choose, I think we should re-export the trait directly from lib.rs. So that people can do use rand::SliceX. No need to expose them to the internal module I think.

It seems one cannot publicly export `pub(crate)` items, hence the only way to export hidden items from a public sub-module is via `#[doc(hidden)]`, which I'd prefer not to use as a permanent solution.

Of course this doesn't stop us moving the traits and still exporting from the lib top-level, or exporting from both the top-level and `seq`, but since we also put these in the prelude is that not enough? These traits feel like the should fit in the `seq` module and I don't see a good reason to export them in two places (ignoring the prelude).

dhardy commented:
Rebased and renamed traits.

> thus I guess we could have list.choose(rng) for one elt and list.sample(rng, number)

Is this clear enough? It seems like the best choice for me *so long as it's not confusing*.

sicking commented:
I feel that it's somewhat confusing that "sample" would mean "multiple". Especially when it doesn't for `distribution.sample(rng)`.

But I can totally buy the argument that "pick" doesn't mean "without replacement".

dhardy commented:
When sampling from iterators, generic/weighted sampling could perhaps be the normal option, e.g. by optionally taking a function which maps elements to weights (whether these weights are then internal to the element or defined externally then doesn't much matter). I'm not sure, but this might not have any performance overhead when not used.

When sampling from slices however, the assumption that all elements are equally likely allows us to simply select a number uniformly from the range `0..len`. With weighting a different approach is required (perhaps calculating cumulative weights then selecting an element with binary search). At a minimum this requires a different selection method depending on whether weighting is used and also implies a separate set-up step may be useful, hence a different API is appropriate.

This correlates with my personal experience: probably every time I have used sampling *with replacement*, I wanted some kind of distribution object with a set-up step and a sampling step returning a single sample, with or without weighting of elements.

(Weighted sampling without replacement is technically possible but I've never found a use for it, and if it was used I would imagine the best API would still be (1) a set-up method and (2) a sample-one-item method, only sampling should remove the item from the collection.)

dhardy commented:
Looking at languages / libraries I only found the following (most languages don't seem to have this functionality within a standard library):

- Python supports `random.choice(seq)`, `random.choices(population, weights=.., cum_weights=.., k)` (normal *or* cumulative weights can be used), `random.sample(population, k)`
- GSL has `gsl_ran_choose(rng, dest, k, src, n, size)` (sampling without replacement) and `gsl_ran_sample(rng, dest, k, src, n, size)` (sampling with replacement)

In both cases, `k` is number of samples to take. GSL uses `n` for the number of elements in the source array and `size` for the size of elements.

We normally use verbs for functions (`choose`) rather than nouns (`choice`). Perhaps we should follow GSL's convention, at least in using the word `choose` for sampling-without-replacement (also used in "n choose k"; see [Binomial coefficient](https://en.wikipedia.org/wiki/Binomial_coefficient)), but we (probably?) still want both single and multiple versions. So:

1. `choose` and `choose_multiple`
2. `choose` and `choose_k`
3. `choose_single` and `choose`
4. `choose_1` and `choose`
5. `choose` (multiple variant only)

Thoughts? I prefer 1, 2 and 5 over the other options.

vks commented:
What is the motivation for the variant that chooses one element? I don't see how `choose_1(&mut rng)` is preferable over `choose(&mut rng, 1)`, so I would prefer option 5 over the others.

dhardy commented:
There is some motivation for the single-element version: (a) simpler, (b) `Option<T>` is nicer to use than `[T]` of length 0 or 1 and (c) doesn't require an allocator.

I suppose we could add `SliceRandom::choose_multiple_fill` using only Floyd's algorithm for sampling; i.e. it wouldn't need an allocator but would be slow for large numbers of elements (more than 400-500 samples; also not optimal when the sample space is small relative to the sample size).

sicking commented:
One thing to keep in mind is that @pitdicker has a different implementation of `choose_multiple` over in https://github.com/pitdicker/rand/blob/slice_random/src/seq.rs, which looks like it doesn't require allocation.

I don't know if there's a meaningful performance difference between its `choose_multiple(1)` and our current `choose` though.

I too have a preference for 1, 2, and 5 (though don't have an opinion about if we use "choose", "pick" or "sample" as base-word).

One important aspect is that it might not be worth having APIs for both "choose one" and "choose multiple" if neither is a common operation. Searching github I see 7K hits for "gen_range", almost all of which are calls to rand. Searching for "choose" gives 5K hits, of which below 1/10 are calls to rand. So `gen_range` is an order of magnitude more used than `choose`.

I think I like the option to just add a `choose_k` method for now. That keeps the name clear but still not too annoying even if you just want one item (whereas `choose_multiple(1)` arguably is a little awkward). And if people complain, or we see a lot more usage, it'd be a non-breaking change to add `choose` later.

dhardy commented:
A quick benchmark shows @pitdicker's `pick_multiple` has its merits:
```
test seq_shuffle_100                     ... FAILED
test seq_slice_choose_multiple_10_of_100 ... bench:         561 ns/iter (+/- 21)
test seq_slice_choose_multiple_1k_of_1M  ... bench:      50,009 ns/iter (+/- 16,255)
```
vs what's in this branch:
```
test seq_shuffle_100                     ... bench:       1,028 ns/iter (+/- 23)
test seq_slice_choose_multiple_10_of_100 ... bench:         147 ns/iter (+/- 2)
test seq_slice_choose_multiple_1k_of_1M  ... bench:      67,401 ns/iter (+/- 4,221)
```
(Both exclude my sample-indices optimisations in #479, though for these bounds it shouldn't matter much.)

dhardy commented:
Re-run, with two different ways of choosing 1 of 1000:
```
# pitdicker's branch
test seq_slice_choose_1_of_1000          ... bench:           5 ns/iter (+/- 0)
test seq_slice_choose_multiple_10_of_100 ... bench:         562 ns/iter (+/- 14)
test seq_slice_choose_multiple_1_of_1000 ... bench:          56 ns/iter (+/- 3)
test seq_slice_choose_multiple_1k_of_1M  ... bench:      48,229 ns/iter (+/- 5,231)

# this branch
test seq_slice_choose_1_of_1000              ... bench:           4 ns/iter (+/- 0)
test seq_slice_choose_multiple_10_of_100     ... bench:         146 ns/iter (+/- 1)
test seq_slice_choose_multiple_1_of_1000     ... bench:          78 ns/iter (+/- 5)
```
So I think there's good reason to have a single-sample choose method (#479 may help but definitely not this much).

dhardy commented:
I cleaned up this branch, squashing most of the fixes.

`IteratorRandom::choose` now uses `gen_bool` instead of `gen_ratio` because casting `usize` → `u32` → `f64` doesn't make much sense. I also removed the checks, because if you're iterating over 2^53 elements then accuracy isn't going to be your biggest problem!

I kept the method names `choose` and `choose_multiple` for now. The brevity of `choose_k` is nice, but I don't know if it would be obvious enough why we use `k`.

It would be nice to get this merged now, even if some details might still change. Any more comments before doing so? I'm happy to wait a day or two for reviews.

My observations:

- #479 has almost no overlap with these changes but does improve performance
- #490 has some overlap and some additional features: better tests and weighted sampling
- [pitdicker/slice_random](https://github.com/pitdicker/rand/blob/slice_random/src/seq.rs) has an alternative approach to `choose_multiple` which is sometimes faster, sometimes slower (see above); we may be able to build some optimisations from this and/or can try building an allocation-free alternative (although probably we shouldn't worry much about allocation)

sicking commented on src/seq.rs:
```diff
@@ -9,62 +9,354 @@
+    /// This is an efficient method to select `amount` elements at random from
+    /// the slice, provided the slice may be mutated.
+    ///
+    /// If you only need to chose elements randomly and `amount > self.len()/2`
```
"chose" -> "choose"

sicking commented on src/seq.rs:
```diff
@@ -9,62 +9,354 @@
+        for (i, elem) in self.enumerate() {
+            let k = rng.gen_range(0, i + 1 + amount);
+            if k < amount {
+                buf[k] = elem;
```
We should be consistent about how we do this check. Right now `choose_multiple_fill` does `if k < amount` whereas `choose_multiple` does `if let Some(spot) = reservoir.get_mut(k) {`. I'd tend to using the latter but either is fine as long as we're consistent.

dhardy commented on src/seq.rs:
```diff
@@ -9,62 +9,354 @@
+/// 
+/// An implementation is provided for slices. This may also be implementable for
+/// other types.
+pub trait SliceRandom {
```
The prelude is not supposed to be the primary place to expose any functionality. I don't really see a reason not to expose this name here.

dhardy commented on benches/seq.rs:
```diff
@@ -0,0 +1,99 @@
+}
+
+#[bench]
+fn seq_iter_choose_multiple_10_of_100(b: &mut Bencher) {
```
I `cfg`'d the whole module and tested (`cargo +nightly test --benches --no-default-features [--features=alloc]`). The bit I don't get is why there are no error messages with `alloc` without `std`, but there aren't.

dhardy replied:
The `cfg` thing made the tests work by effectively removing all the benchmarks — took me a while to work out why no benchmarks were getting run. But we don't need to benchmark without `std` anyway.

dhardy commented on benches/seq.rs:
```diff
@@ -0,0 +1,99 @@
+    let mut buf = [0; 1];
+    b.iter(|| {
+        for (v, slot) in x.choose_multiple(&mut rng, buf.len()).zip(buf.iter_mut()) {
+            *slot = *v;
```
Just calling `next()` doesn't *do* anything but return an iterator shim designed to be optimised into something else. Maybe `.cloned().next()`...

dhardy replied:
`.cloned()` should make it return a copy.

Anyway, this change appears to change performance from 82ns to 79ns (higher baseline because I didn't merge other branch here).

dhardy commented on src/seq.rs:
```diff
@@ -9,62 +9,354 @@
+    ///
+    /// Depending on the implementation, complexity is expected to be `O(m)`,
+    /// where `m = amount`.
+    fn partial_shuffle<R>(&mut self, rng: &mut R, amount: usize)
```
If we mention this here it becomes part of the function specification, i.e. the implementation can't change it in case it breaks users' assumptions.

Okay for complexity comment.

sicking commented on src/seq.rs:
```diff
@@ -9,62 +9,354 @@
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // TODO: investigate using SliceIndex::get_unchecked when stable
+        self.indices.next().map(|i| &(*self.slice)[i])
```
`slice::get_unchecked` looks like it has been marked stable since 1.0?
https://doc.rust-lang.org/src/alloc/slice.rs.html#419-423

dhardy replied:
Yes, but all we know about `self.slice` is the type bounds: `'a, S: Index<usize, Output = T> + ?Sized + 'a, T: 'a`. We need to use the `SliceIndex` trait, which apparently is going to be marked stable soon, but with most of the methods still unstable.

dhardy commented on benches/seq.rs:
```diff
@@ -0,0 +1,99 @@
+    let mut buf = [0; 10];
+    b.iter(|| {
+        for (v, slot) in x.choose_multiple(&mut rng, buf.len()).zip(buf.iter_mut()) {
+            *slot = *v;
```
I don't think this is significant. (I have tested that writing to a buffer is faster than collecting into a vector.)

dhardy commented:
Updated (rebase, but only minor tweaks)
